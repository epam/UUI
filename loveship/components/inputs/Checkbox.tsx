import * as types from '../types';
import css from './Checkbox.scss';
import { Checkbox as uuiCheckbox, CheckboxProps } from '@epam/uui-components';
import { withMods } from '@epam/uui-core';
import { ReactComponent as TickIcon } from '../icons/checkbox-checked.svg';
import { ReactComponent as IndeterminateIcon } from '../icons/checkbox-partial.svg';
import styles from '../../assets/styles/scss/loveship-color-vars.scss';

export interface CheckboxMods extends types.ColorMod {
    size?: '12' | '18';
    theme?: 'light' | 'dark';
    mode?: 'form' | 'cell';
}

export function applyCheckboxMods(mods: CheckboxMods & CheckboxProps) {
    return [
        css.root,
        css['size-' + (mods.size || '18')],
        css['theme-' + (mods.theme || 'light')],
        css['mode-' + (mods.mode || 'form')],
        styles['color-' + (mods.color || 'sky')],
    ];
}

export const Checkbox = withMods<CheckboxProps, CheckboxMods>(uuiCheckbox, applyCheckboxMods, () => ({
    icon: TickIcon,
    indeterminateIcon: IndeterminateIcon,
}));
