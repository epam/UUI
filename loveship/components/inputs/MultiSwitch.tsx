import React from 'react';
import { IEditable, IHasRawProps } from '@epam/uui';
import { ButtonProps } from '@epam/uui-components';
import cx from 'classnames';
import { Button, ControlGroup } from '../index';
import * as types from '../types';
import * as css from './MultiSwitch.scss';

interface MultiSwitchItem<TValue> extends ButtonProps, types.ColorMod {
    id: TValue;
}

export interface MultiSwitchProps<TValue> extends IEditable<TValue>, types.ColorMod, types.SizeMod, IHasRawProps<HTMLDivElement> {
    items: MultiSwitchItem<TValue>[];
}

export class MultiSwitch<TValue> extends React.Component<MultiSwitchProps<TValue>, any> {

    render() {
        return (
            <ControlGroup rawProps={ { ...this.props.rawProps, role: 'tablist' } }>
                {
                    this.props.items.map((item, index) => {
                        const isActive = this.props.value === item.id;

                        return (
                            <Button
                                {  ...item }
                                cx={ cx(item.cx, { [css.selectedItem]: isActive }) }
                                key={ index + '-' + item.id as any }
                                onClick={ () => this.props.onValueChange(item.id) }
                                shape="square"
                                fill={ isActive ? 'solid' : 'white' }
                                color={ item.color || this.props.color }
                                size={ this.props.size }
                                rawProps={{ 'aria-current': this.props.value === item.id, role: 'tab' }}
                            />
                        );
                    })
                }
            </ControlGroup>
        );
    }
}
