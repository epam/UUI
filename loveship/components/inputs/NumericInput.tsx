import * as types from '../types';
import css from './NumericInput.scss';
import colorStyle from '../../assets/styles/scss/loveship-color-vars.scss';
import { withMods } from '@epam/uui-core';
import { NumericInput as uuiNumericInput, NumericInputProps } from '@epam/uui-components';
import { TextSettings, getTextClasses } from '../../helpers/textLayout';
import { systemIcons } from '../icons/icons';

const defaultSize = '36';

export interface NumericInputMods extends types.EditMode, types.SizeMod, TextSettings {}

export function applyNumericInputMods(mods: NumericInputMods) {
    return [colorStyle.colorSky, css.root, css['size-' + (mods.size || defaultSize)], css['mode-' + (mods.mode || 'form')]];
}

export const NumericInput = withMods<NumericInputProps, NumericInputMods>(uuiNumericInput, applyNumericInputMods, props => ({
    upIcon: systemIcons[props.size || defaultSize].foldingArrow,
    downIcon: systemIcons[props.size || defaultSize].foldingArrow,
    inputCx: getTextClasses(
        {
            size: props.size || defaultSize,
            lineHeight: props.lineHeight,
            fontSize: props.fontSize,
        },
        true
    ),
    align: props.align ?? (props.mode === 'cell' ? 'right' : 'left'),
    disableArrows: props.disableArrows ?? props.mode === 'cell',
}));
