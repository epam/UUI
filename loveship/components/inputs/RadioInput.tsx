import css from './RadioInput.scss';
import styles from '../../assets/styles/scss/loveship-color-vars.scss';
import { RadioInput as uuiRadioInput, RadioInputProps } from '@epam/uui-components';
import { withMods } from '@epam/uui-core';
import { ReactComponent as RadioPoint } from '../icons/radio-point.svg';
import * as types from '../types';

export interface RadioInputMods {
    size?: '12' | '18';
    color?: types.ColorMod;
    theme?: 'light' | 'dark';
}

function applyRadioInputMods(mods: RadioInputMods & RadioInputProps) {
    return [css.root, css['size-' + (mods.size || '18')], styles['color-' + (mods.color || 'sky')], css['theme-' + (mods.theme || 'light')]];
}

export const RadioInput = withMods<RadioInputProps, RadioInputMods>(uuiRadioInput, applyRadioInputMods, () => ({ icon: RadioPoint }));
