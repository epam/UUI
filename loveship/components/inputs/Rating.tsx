import { withMods } from '@epam/uui-core';
import * as uuiComponents from '@epam/uui-components';
import css from './Rating.module.scss';
import { ReactComponent as FilledStarIcon } from '../icons/star-filled.svg';
import { ReactComponent as EmptyStarIcon } from '../icons/star-empty.svg';
import { Tooltip } from '../overlays';

export interface RatingMods {
    /**
     * @default '18'
     */
    size?: 18 | 24 | 30;
}
export interface RatingProps extends uuiComponents.RatingProps, RatingMods {}

function applyRatingMods(mods: RatingMods & uuiComponents.RatingProps) {
    return [
        css.root,
        css['size-' + (mods.size || '18')],
    ];
}

export const Rating = withMods<uuiComponents.RatingProps, RatingMods>(
    uuiComponents.Rating,
    applyRatingMods,
    () => ({ filledStarIcon: FilledStarIcon, emptyStarIcon: EmptyStarIcon, Tooltip }),
);
