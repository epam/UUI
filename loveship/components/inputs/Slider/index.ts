export * from './RangeSlider';
export * from './Slider';
