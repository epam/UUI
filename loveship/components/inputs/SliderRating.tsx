import React, { useState, useRef } from 'react';
import { BaseRating, IconContainer } from '@epam/uui-components';
import { Icon, IEditable, IHasRawProps } from '@epam/uui-core';
import css from './SliderRating.module.scss';
import { ReactComponent as LineGrayIcon } from '../icons/slider-rating/line_gray_icon.svg';
import { ReactComponent as LineRedIcon } from '../icons/slider-rating/line_red_icon.svg';
import { ReactComponent as LineYellowIcon } from '../icons/slider-rating/line_yellow_icon.svg';
import { ReactComponent as LineBlueIcon } from '../icons/slider-rating/line_blue_icon.svg';
import { ReactComponent as LineVioletIcon } from '../icons/slider-rating/line_violet_icon.svg';
import { ReactComponent as ActiveMarkRedIcon } from '../icons/slider-rating/active_mark_red_icon.svg';
import { ReactComponent as ActiveMarkYellowIcon } from '../icons/slider-rating/active_mark_yellow_icon.svg';
import { ReactComponent as ActiveMarkGreenIcon } from '../icons/slider-rating/active_mark_green_icon.svg';
import { ReactComponent as ActiveMarkBlueIcon } from '../icons/slider-rating/active_mark_blue_icon.svg';
import { ReactComponent as ActiveMarkVioletIcon } from '../icons/slider-rating/active_mark_violet_icon.svg';
import { ReactComponent as NaIcon } from '../icons/slider-rating/na_icon.svg';
import { ReactComponent as NaActiveIcon } from '../icons/slider-rating/na_active_icon.svg';
import { Tooltip } from '../overlays';
import cx from 'classnames';
import { i18n } from '../../i18n';

const defaultSize = '18';

export interface SliderRatingProps<TValue> extends IEditable<TValue>, IHasRawProps<React.HTMLAttributes<HTMLDivElement>> {
    renderTooltip?: (value: TValue) => React.ReactNode;
    /**
     * @default 1
     */
    from?: 1 | 2;
    /**
     * @default false
     */
    withoutNa?: boolean;
    /**
     * @default '18'
     */
    size?: '18' | '24';
    getScaleIcon?: (value: number) => Icon;
    getHandlerIcon?: (value: number) => Icon;
    tooltipColor?: 'white' | 'fire' | 'gray';
}

const maxValue = 5;

export class SliderRating extends React.Component<SliderRatingProps<number>> {
    handlerWidth: number;
    getScaleIcon = (rating: number) => {
        switch (rating) {
            case 1:
                return LineRedIcon;
            case 2:
                return LineYellowIcon;
            case 3:
                return LineGrayIcon;
            case 4:
                return LineBlueIcon;
            case 5:
                return LineVioletIcon;
            default:
                return LineGrayIcon;
        }
    };

    getHandlerIcon = (rating: number) => {
        switch (rating) {
            case 1:
                return ActiveMarkRedIcon;
            case 2:
                return ActiveMarkYellowIcon;
            case 3:
                return ActiveMarkGreenIcon;
            case 4:
                return ActiveMarkBlueIcon;
            case 5:
                return ActiveMarkVioletIcon;
        }
    };

    getLeftHandlerIconPosition = (rating: number, from: number, stepWidth: number) => {
        const left = !!rating ? (rating - from) * stepWidth - this.handlerWidth / 2 : 0;

        if (rating && rating === from) {
            return left + 2;
        } else if (rating && rating === maxValue) {
            return left - 2;
        }

        return left;
    };

    renderTooltipBox(rating: number) {
        const tooltipContent = this.props.renderTooltip ? this.props.renderTooltip(rating) : `${rating}`;
        return <TooltipBox tooltipColor={ this.props.tooltipColor } content={ tooltipContent } size={ this.props.size || '18' } />;
    }

    renderRating = (sliderRating: number, markWidth: number, numberOfMarks: number) => {
        const rating = sliderRating || 0;
        const from = this.props.from || 1;
        const stepWidth = (markWidth * numberOfMarks) / (numberOfMarks - 1);
        const left = this.getLeftHandlerIconPosition(rating, from, stepWidth);
        const size = this.props.size || defaultSize;

        return (
            <>
                <div className={ cx(css.scale, css[`size-${size}`], from === 2 && css.shortScale) }>
                    <IconContainer cx={ css.scaleIcon } icon={ this.props.getScaleIcon ? this.props.getScaleIcon(rating) : this.getScaleIcon(rating) } />
                </div>
                {this.renderTooltipBox(rating)}
                <div
                    className={ cx(css.handler, css[`size-${size}`], !rating && css.hidden) }
                    style={ { left: left } }
                    ref={ (handler) => {
                        this.handlerWidth = handler && handler.offsetWidth;
                    } }
                >
                    <Tooltip color={ this.props.tooltipColor } cx={ css.tooltip } content={ this.props.renderTooltip ? this.props.renderTooltip(rating) : `${rating}` }>
                        <IconContainer cx={ css.handlerIcon } icon={ this.props.getHandlerIcon ? this.props.getHandlerIcon(rating) : this.getHandlerIcon(rating) } />
                    </Tooltip>
                </div>
            </>
        );
    };

    renderNa() {
        const isReadonly = this.props.isReadonly || this.props.isDisabled;
        const size = this.props.size || defaultSize;

        if (isReadonly && this.props.value !== 0) {
            return <IconContainer cx={ cx(css.naIcon, css[`size-${size}`], css.disabled) } icon={ NaIcon } />;
        } else {
            return (
                <Tooltip color={ this.props.tooltipColor } content={ this.props.renderTooltip ? this.props.renderTooltip(0) : i18n.sliderRating.notAvailableMessage }>
                    <IconContainer
                        cx={ cx(css.naIcon, css[`size-${size}`], isReadonly && css.disabled) }
                        icon={ this.props.value === 0 ? NaActiveIcon : NaIcon }
                        onClick={ !isReadonly && (() => this.props.onValueChange(0)) }
                    />
                </Tooltip>
            );
        }
    }

    render() {
        return (
            <div className={ css.container } { ...this.props.rawProps }>
                <BaseRating
                    from={ this.props.from || 1 }
                    to={ maxValue }
                    step={ 1 }
                    renderRating={ this.renderRating }
                    { ...this.props }
                    value={ this.props.value === 0 ? null : this.props.value }
                    cx={ css.baseRatingContainer }
                />
                {!this.props.withoutNa && <div className={ css.naIconContainer }>{this.renderNa()}</div>}
            </div>
        );
    }
}

type TooltipBoxProps = {
    size: string;
    tooltipColor: 'white' | 'fire' | 'gray';
    content: React.ReactNode | string;
};

function TooltipBox(props: TooltipBoxProps) {
    const { content, size, tooltipColor } = props;
    const tooltipBoxRef = useRef<HTMLDivElement>(null);
    const [left, setLeft] = useState<number>(0);

    const topPosition = tooltipBoxRef.current?.getBoundingClientRect().y || 0;

    return (
        <div className={ css.tooltipsBox } ref={ tooltipBoxRef } onMouseMove={ (event: React.MouseEvent) => setLeft(event.clientX) }>
            <Tooltip color={ tooltipColor } placement="top" content={ content } cx={ css.tooltip }>
                <div
                    className={ css.tooltipsBoxItem }
                    style={ {
                        left: `${left - 1}px`,
                        top: `${topPosition}px`,
                        height: `${size}px`,
                    } }
                />
            </Tooltip>
            <Tooltip color={ tooltipColor } placement="top" content={ content } cx={ css.tooltip }>
                <div
                    className={ css.tooltipsBoxItem }
                    style={ {
                        left: `${left}px`,
                        top: `${topPosition}px`,
                        height: `${size}px`,
                    } }
                />
            </Tooltip>
            <Tooltip color={ tooltipColor } placement="top" content={ content } cx={ css.tooltip }>
                <div
                    className={ css.tooltipsBoxItem }
                    style={ {
                        left: `${left + 1}px`,
                        top: `${topPosition}px`,
                        height: `${size}px`,
                    } }
                />
            </Tooltip>
        </div>
    );
}
