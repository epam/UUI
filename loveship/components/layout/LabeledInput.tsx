import css from './LabeledInput.scss';
import styles from '../../assets/styles/scss/loveship-color-vars.scss';
import * as types from '../types';
import { withMods } from '@epam/uui-core';
import { LabeledInput as uuiLabeledInput, LabeledInputProps } from '@epam/uui-components';
import { Tooltip } from '../overlays';
import { systemIcons } from '../icons/icons';

const defaultSize = '36';

export interface LabeledInputMods extends types.ColorMod, types.SizeMod {}

function applyLabeledInputMods(mods: LabeledInputMods) {
    return [css.root, css['size-' + (mods.size || defaultSize)], styles['color-' + (mods.color || 'night700')]];
}

export const LabeledInput = withMods<LabeledInputProps, LabeledInputMods>(uuiLabeledInput, applyLabeledInputMods, props => ({
    Tooltip,
    infoIcon: systemIcons[props.size || defaultSize].info,
}));
