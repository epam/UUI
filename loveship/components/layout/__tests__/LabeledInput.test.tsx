import React from "react";
import { renderSnapshotWithContextAsync } from '@epam/test-utils';
import { LabeledInput } from "../LabeledInput";
import { TextInput } from "../../inputs";

describe('LabeledInput', () => {
    it('should be rendered correctly', async () => {
        const tree = await renderSnapshotWithContextAsync(
            <LabeledInput>
                <TextInput value='test' onValueChange={ jest.fn() } />
            </LabeledInput>
        );

        expect(tree).toMatchSnapshot();
    });

    it('should be rendered correctly with props', async () => {
        const tree = await renderSnapshotWithContextAsync(
            <LabeledInput
                label='Test label'
                size='36'
                info='Test'
                isInvalid
                validationMessage='Test invalid message'
                labelPosition='left'
            >
                <TextInput value='test'  onValueChange={ jest.fn() } />
            </LabeledInput>
        );

        expect(tree).toMatchSnapshot();
    });
});
