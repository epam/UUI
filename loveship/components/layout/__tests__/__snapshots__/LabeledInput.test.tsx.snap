// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LabeledInput should be rendered correctly 1`] = `
<div
  className="container root size-36 color-night700"
>
  <div
    className="uui-label-top"
  >
    <div>
      <div
        className="container uui-input-box -clickable colorSky root size-36 mode-form"
        onBlur={[Function]}
        onFocus={[Function]}
        tabIndex={-1}
      >
        <input
          className="uui-input line-height-18 font-size-14 v-padding-8"
          onChange={[Function]}
          onKeyDown={[Function]}
          type="text"
          value="test"
        />
      </div>
    </div>
  </div>
</div>
`;

exports[`LabeledInput should be rendered correctly with props 1`] = `
<div
  className="container root size-36 color-night700"
>
  <div
    className="uui-label-left"
  >
    <div
      className="uui-label"
    >
      Test label
      <svg
        className="uui-labeled-input-info-icon"
        height={24}
        viewBox="0 0 12 24"
        width={12}
      >
        <use
          xlinkHref="#info-18.svg"
        />
      </svg>
    </div>
    <div
      className="rightChildrenPosition"
    >
      <div
        className="container uui-input-box -clickable colorSky root size-36 mode-form"
        onBlur={[Function]}
        onFocus={[Function]}
        tabIndex={-1}
      >
        <input
          className="uui-input line-height-18 font-size-14 v-padding-8"
          onChange={[Function]}
          onKeyDown={[Function]}
          type="text"
          value="test"
        />
      </div>
    </div>
  </div>
  <div
    className="uui-invalid-message"
  >
    Test invalid message
  </div>
</div>
`;
